backend/src/v2/driver/driver.go

```diff
@@ -1927,61 +1927,83 @@ func resolveUpstreamParameters(cfg resolveUpstreamOutputsConfig) (*structpb.Valu
 func resolveUpstreamArtifacts(cfg resolveUpstreamOutputsConfig) (*pipelinespec.ArtifactList, error) {
 	glog.V(4).Infof("artifactSpec: %#v", cfg.artifactSpec)
 	taskOutput := cfg.artifactSpec.GetTaskOutputArtifact()
+	glog.V(4).Info("taskOutput: ", taskOutput)
+	producerTaskName := taskOutput.GetProducerTask()
 	if taskOutput.GetProducerTask() == "" {
 		return nil, cfg.err(fmt.Errorf("producer task is empty"))
 	}
 	if taskOutput.GetOutputArtifactKey() == "" {
 		cfg.err(fmt.Errorf("output artifact key is empty"))
 	}
+	producerTaskName = metadata.UpdateTaskNameWithDagID(producerTaskName, cfg.dag.Execution.GetID())
+	// The main difference between the root ParallelFor DAG and its iteration
+	// DAGs is that the root contains the customer property "iteration_count"
+	// while the iterations contain "iteration_index". We can use this to
+	// determine if we are in a ParallelFor DAG or not. The iteration DAGs will
+	// contain the "iteration_index" which is used to resolve the correct output
+	// artifact for the downstream task within the iteration. ParallelFor
+	// iterations are DAGs themselves, we can verify if we are in a iteration by
+	// confirming that the "iteration_index" exists for the DAG of the current
+	// task we are attempting to resolve. If the dag contains the
+	// "iteration_index", the producerTaskName will be updated appropriately
+	producerTaskName = InferIndexedTaskName(producerTaskName, cfg.dag.Execution)
+	glog.V(4).Infof("producerTaskName: %v", producerTaskName)
 	tasks, err := getDAGTasks(cfg.ctx, cfg.dag, cfg.pipeline, cfg.mlmd, nil)
 	if err != nil {
-		cfg.err(err)
+		return nil, cfg.err(err)
 	}
 
-	producer, ok := tasks[taskOutput.GetProducerTask()]
+	producer, ok := tasks[producerTaskName]
 	if !ok {
-		cfg.err(
-			fmt.Errorf("cannot find producer task %q", taskOutput.GetProducerTask()),
+		return nil, cfg.err(
+			fmt.Errorf("cannot find producer task %q", producerTaskName),
 		)
 	}
 	glog.V(4).Info("producer: ", producer)
+	glog.V(4).Infof("tasks: %#v", tasks)
 	currentTask := producer
 	outputArtifactKey := taskOutput.GetOutputArtifactKey()
 
-	// Continue looping until we reach a sub-task that is NOT a DAG.
+	// Continue looping until we reach a sub-task that is either a ParallelFor task or a Container task.
 	for {
 		glog.V(4).Info("currentTask: ", currentTask.TaskName())
 		// If the current task is a DAG:
 		if *currentTask.GetExecution().Type == "system.DAGExecution" {
```

Comment:
Could there be cases in this conditional where it's a DAG but not a ParallelFor?